### PR TITLE
Generate `ImportMap` from module path to imported dependencies

### PR DESCRIPTION
**Pull request by chanman3388**
Related to #2914 amongst others.
The tests will fail at the moment but I'd like to get some feedback on the architecture first.
So I've gone the route of making a separate struct `ast::types::Import` to hold the information about an import. This includes expanding `ImportFrom` statements using `checkers::imports::check_import`.
In addition to producing the diagnostics, this now returns a tuple of the diagnostics and a hashmap with an `Option<PathBuf>` as the key (will this ever be `None`?), and the value as a `Vec` of these new `Import` structs.
Eventually one can retrieve a `Vec` of these hashmaps inside of `run.rs`.
I haven't yet taken into account wildcard imports, or if a module has an `__all__` which exports the modules under different names.

### CODE REVIEW

**Comment by chanman3388** [comment]
So now I have a hashmap of the paths to a vec of the imports they require like so:
```
[2023-03-12][18:21:35][ruff_cli::commands::run][DEBUG] {
    "/home/chris/projects/pylint_exp/R0401/src/a.py": [
        Import {
            name: ".b",
            location: Location {
                row: 1,
                column: 14,
            },
            end_location: Location {
                row: 1,
                column: 15,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/f/a.py": [
        Import {
            name: "src.g.a",
            location: Location {
                row: 1,
                column: 18,
            },
            end_location: Location {
                row: 1,
                column: 24,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/f/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/g/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/j/a.py": [
        Import {
            name: "src.a",
            location: Location {
                row: 1,
                column: 16,
            },
            end_location: Location {
                row: 1,
                column: 17,
            },
        },
        Import {
            name: "src.b",
            location: Location {
                row: 1,
                column: 19,
            },
            end_location: Location {
                row: 1,
                column: 20,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/c/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/g/a.py": [
        Import {
            name: "src.e.a",
            location: Location {
                row: 1,
                column: 0,
            },
            end_location: Location {
                row: 1,
                column: 14,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/d/a.py": [
        Import {
            name: "src.c.a",
            location: Location {
                row: 1,
                column: 0,
            },
            end_location: Location {
                row: 1,
                column: 14,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/j/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/b.py": [
        Import {
            name: ".a",
            location: Location {
                row: 1,
                column: 14,
            },
            end_location: Location {
                row: 1,
                column: 15,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/e/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/d/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/c/a.py": [
        Import {
            name: "src.d.a",
            location: Location {
                row: 1,
                column: 18,
            },
            end_location: Location {
                row: 1,
                column: 19,
            },
        },
    ],
    "/home/chris/projects/pylint_exp/R0401/src/__init__.py": [],
    "/home/chris/projects/pylint_exp/R0401/src/e/a.py": [
        Import {
            name: "src.f.a",
            location: Location {
                row: 1,
                column: 0,
            },
            end_location: Location {
                row: 1,
                column: 14,
            },
        },
    ],
}
```

For a file structure like this:
```
src
├── __init__.py
├── a.py
├── b.py
├── c
│   ├── __init__.py
│   └── a.py
├── d
│   ├── __init__.py
│   └── a.py
├── e
│   ├── __init__.py
│   └── a.py
├── f
│   ├── __init__.py
│   └── a.py
├── g
│   ├── __init__.py
│   └── a.py
└── j
    ├── __init__.py
    └── a.py
```

**Comment by chanman3388** [comment]
TODO: Still need to figure out a way to expand relative imports

**Comment by charliermarsh** [comment]
Great! I'll try to give this a thorough review tomorrow.

File: crates/ruff/src/checkers/imports.rs
Comment (MichaReiser) [inline crates/ruff/src/checkers/imports.rs:29]: `FxHashMap<PathBuf, Vec<Import>>` is now used in many places. Can we introduce a new-type that gives it a name and abstracts a way the internal data structure

```rust
#[derive(Debug, Clone, Default)]
struct Imports {
	inner: FxHashMap<PathBuf, Vec<Import>>
}

impl Imports {
	pub fn insert(...) {}
}
```

File: crates/ruff/src/checkers/imports.rs
Comment (MichaReiser) [inline crates/ruff/src/checkers/imports.rs:67]: Not specific to this PR: @charliermarsh it seems that it is guaranteed that location is always set. Can we introduce a helper function / extension method that calls the `unwrap` rather than sprinkling the unwraps all over our code base?

File: crates/ruff/src/checkers/imports.rs
Comment (MichaReiser) [inline crates/ruff/src/checkers/imports.rs:81]: Nit: Does this work
```suggestion
                                module.unwrap_or_default()
```

File: crates/ruff/src/checkers/imports.rs
Comment (MichaReiser) [inline crates/ruff/src/checkers/imports.rs:88]: Is it necessary to call `collect` here? `extend` should be fine with an `Iter` argument
```suggestion
```

File: crates/ruff/src/checkers/imports.rs
Comment (MichaReiser) [inline crates/ruff/src/checkers/imports.rs:90]: Nit: Can we add a small comment explaining what guarantees that there should only be import statements?

File: crates/ruff/src/checkers/imports.rs
Comment (MichaReiser) [inline crates/ruff/src/checkers/imports.rs:95]:
```suggestion
```

File: crates/ruff/src/linter.rs
Comment (MichaReiser) [inline crates/ruff/src/linter.rs:70]: Can you explain why the lifetime annotation is necessary? We aren't using it in any return type.

File: crates/ruff/src/linter.rs
Comment (MichaReiser) [inline crates/ruff/src/linter.rs:55]: These type seem to be somewhat common too. It could make sense to wrap them in a newtype-wrapper to avoid exposing internals (and it gives you the possibility to implement additional methods on them)

File: crates/ruff_cli/src/commands/run.rs
Comment (MichaReiser) [inline crates/ruff_cli/src/commands/run.rs:135]: Is this a todo?

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:29]: Sure, I'll look into doing this

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:81]: I'll give it a go

File: crates/ruff/src/linter.rs
Comment (chanman3388) [reply crates/ruff/src/linter.rs:70]: I think this is probably a hangover from when I was trying to make the imports work on a `&Path` rather than a `PathBuf`, so yes I will need to get rid of these

File: crates/ruff_cli/src/commands/run.rs
Comment (chanman3388) [reply crates/ruff_cli/src/commands/run.rs:135]: Yes it is, I'll edit the comment appropriately

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:81]: Hm, I think that as the `Default` trait isn't implement for `std::string::String` we can't _just_ do this, but I can do something similar.

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:88]: I'll have a look

Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:88]: Forgive me for being a noob, but if we don't return a `Vec` here, won't this get a bit more complicated as then the hashmap has to contain something which will `impl` a trait? As a dynamically sized object, at least when I try to do this, I get the complaint that the size needs to be known at compile time.
Unless you mean that you'd like that, and I need to `Box` it?

File: crates/ruff/src/linter.rs
Comment (MichaReiser) [reply crates/ruff/src/linter.rs]: Up to you. My tendency is to at least wrap types that are public.

File: crates/ruff_cli/src/commands/run.rs
Comment (chanman3388) [inline crates/ruff_cli/src/commands/run.rs:136]: Will get rid of this once everything is good.

**Review by MichaReiser — APPROVED** [review]
LGTM. Thank you for working on this. I'll leave this to @charliermarsh for merging because I'm still learning the code base.

File: crates/ruff/src/checkers/ast/mod.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/ast/mod.rs:2297]: Good spot, I think this was actually for another piece of functionality I was looking to implement. Maybe for `assignment-from-none`? At any rate, it's not meant to be here, I'll get rid of it.

**Comment by chanman3388** [comment]
With regards to relative imports, if I have the time I'll try have a bash at this tomorrow, I think I can use the `level` information from the import to derive what the full path would be for relative import lines.

**Comment by chanman3388** [comment]
So **hopefully** I've manage the relative import thing. Will need to test further, but for now if I have the following file structure:
```
grand
├── __init__.py
├── a.py (from .parent.child import a; from .parent import a; from parent.child import b)
└── parent
    ├── __init__.py
    ├── a.py (from .child import a,b; from .. import a)
    └── child
        ├── __init__.py
        ├── a.py (from ..import a; from ... import a)
        └── b.py (from . import a)
```

I get the following imports out:
```
Imports {
    inner: {
        "/home/chris/projects/pylint_exp/R0401/grand/__init__.py": [],
        "/home/chris/projects/pylint_exp/R0401/grand/parent/__init__.py": [],
        "/home/chris/projects/pylint_exp/R0401/grand/parent/child/__init__.py": [],
        "/home/chris/projects/pylint_exp/R0401/grand/parent/a.py": [
            Import {
                name: "grand.child.a",
                location: Location {
                    row: 2,
                    column: 19,
                },
                end_location: Location {
                    row: 2,
                    column: 20,
                },
            },
            Import {
                name: "grand.child.b",
                location: Location {
                    row: 2,
                    column: 22,
                },
                end_location: Location {
                    row: 2,
                    column: 23,
                },
            },
            Import {
                name: "grand.a",
                location: Location {
                    row: 3,
                    column: 15,
                },
                end_location: Location {
                    row: 3,
                    column: 16,
                },
            },
        ],
        "/home/chris/projects/pylint_exp/R0401/grand/parent/child/a.py": [
            Import {
                name: "grand.a",
                location: Location {
                    row: 2,
                    column: 15,
                },
                end_location: Location {
                    row: 2,
                    column: 16,
                },
            },
            Import {
                name: "grand.a",
                location: Location {
                    row: 3,
                    column: 16,
                },
                end_location: Location {
                    row: 3,
                    column: 17,
                },
            },
        ],
        "/home/chris/projects/pylint_exp/R0401/grand/a.py": [
            Import {
                name: "grand.parent.child.a",
                location: Location {
                    row: 2,
                    column: 26,
                },
                end_location: Location {
                    row: 2,
                    column: 27,
                },
            },
            Import {
                name: "grand.parent.a",
                location: Location {
                    row: 3,
                    column: 20,
                },
                end_location: Location {
                    row: 3,
                    column: 21,
                },
            },
            Import {
                name: "grand.parent.child.b",
                location: Location {
                    row: 4,
                    column: 25,
                },
                end_location: Location {
                    row: 4,
                    column: 26,
                },
            },
        ],
        "/home/chris/projects/pylint_exp/R0401/grand/parent/child/b.py": [
            Import {
                name: "grand.a",
                location: Location {
                    row: 2,
                    column: 14,
                },
                end_location: Location {
                    row: 2,
                    column: 15,
                },
            },
        ],
    },
}
```

**Comment by chanman3388** [comment]
@MichaReiser Sorry I've made some new changes now to deal with resolution of relative imports, would you mind giving it the once over again?

File: crates/ruff/src/checkers/imports.rs
Comment (charliermarsh) [inline crates/ruff/src/checkers/imports.rs:85]: `import os, sys` is valid, so this _can_ have multiple entries.

File: crates/ruff/src/checkers/imports.rs
Comment (charliermarsh) [inline crates/ruff/src/checkers/imports.rs:87]: I think we should introduce lifetimes to avoid the need to clone names. The AST will outlive all of this stuff, so it should be possible to just take references.

File: crates/ruff/src/checkers/imports.rs
Comment (charliermarsh) [inline crates/ruff/src/checkers/imports.rs:81]: Can we reuse `crates/ruff_python_ast/src/helpers.rs#to_module_path` for this?

File: crates/ruff/src/checkers/imports.rs
Comment (charliermarsh) [reply crates/ruff/src/checkers/imports.rs:89]: Yeah for context, the Python AST treats end locations as optional. In RustPython, though, we always provide them (and `Expr::new` _requires_ them). So the _interface_ mirrors CPython, but they _are_ guaranteed in practice.

File: crates/ruff_python_ast/src/types.rs
Comment (charliermarsh) [inline crates/ruff_python_ast/src/types.rs:136]: Can we look at, or reuse, or share logic with `crates/ruff/src/rules/flake8_tidy_imports/relative_imports.rs#fix_banned_relative_import`? I believe it's solving the same problem.

File: crates/ruff_python_ast/src/types.rs
Comment (charliermarsh) [inline crates/ruff_python_ast/src/types.rs:85]: Have we looked ahead to what the needs will be for lint rules that will rely on this, like cyclic imports? It looks like this is a string representation of the module (e.g., `import foo` => `"foo"`). Will we need to translate it to a _file path_ at all? Do we have the information we need about the containing package?

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:87]: I guess that would depend on how we want this to look. If we can make whatever it is we get from relative/from imports to be the same type then it should be doable.

Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:87]: Forgot to mention, the AST does outlive _this_ stuff, but the tokens only exist until the end of the various functions inside of `linter.rs`, if we want the locations and names to live past this they must be owned, at least to the best of my current understanding.

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:81]: I'll look into it.

File: crates/ruff_python_ast/src/types.rs
Comment (chanman3388) [reply crates/ruff_python_ast/src/types.rs:85]: Very happy for others to chime in here, I'm most familiar with pylint as that's what I use on my day job, I ran through the list of items from #970, and honestly I think this is only useful for `cyclic-import R0401`.
One could also use the logic used to report the imports for `import-self (W0406)` I reckon, but that doesn't need knowledge of other imports.

File: crates/ruff/src/checkers/imports.rs
Comment (chanman3388) [reply crates/ruff/src/checkers/imports.rs:81]: My only qualm with using it is that it returns me a `Vec<String>` whereas here we're using `Vec<&str>`

**Comment by charliermarsh** [comment]
High-level question: should we first try to draft the cyclic imports PR that leverages this extracted metadata, prior to merging? Implementing that diagnostic may point out things we need to include here.

**Comment by chanman3388** [comment]
> High-level question: should we first try to draft the cyclic imports PR that leverages this extracted metadata, prior to merging? Implementing that diagnostic may point out things we need to include here.

Seems like a sensible idea.

**Comment by chanman3388** [comment]
So on the grand/parent/child example I have above, pylint throws the following out:
```
************* Module grand.parent.child.__init__
grand/parent/child/__init__.py:1:0: R0401: Cyclic import (grand.parent.a -> grand.parent.child.a) (cyclic-import)
grand/parent/child/__init__.py:1:0: R0401: Cyclic import (grand.a -> grand.parent.child.a -> grand.parent.a) (cyclic-import)
grand/parent/child/__init__.py:1:0: R0401: Cyclic import (grand.a -> grand.parent.child.a) (cyclic-import)
grand/parent/child/__init__.py:1:0: R0401: Cyclic import (grand.a -> grand.parent.a) (cyclic-import)
grand/parent/child/__init__.py:1:0: R0401: Cyclic import (grand.a -> grand.parent.child.b -> grand.parent.child.a -> grand.parent.a) (cyclic-import)
```
I'm still tweaking the cycle detection, it's not quite there yet.

**Comment by chanman3388** [comment]
Interestingly this detects an extra cycle compared to pylint, not sure why this would be the case, I've only briefly looked at pylint's implementation. I haven't yet done the same logic as all the other rules yet, that is to say, it's permanently on.

Here is the example output.
```
grand/a.py:2:1: I001 [*] Import block is un-sorted or un-formatted
grand/a.py:2:27: PLR0401 Cyclic import (grand.a -> grand.parent.child.a -> grand.parent.a)
grand/a.py:2:27: PLR0401 Cyclic import (grand.a -> grand.parent.child.a)
grand/a.py:2:27: PLR0401 Cyclic import (grand.a -> grand.parent.a)
grand/a.py:2:27: PLR0401 Cyclic import (grand.a -> grand.parent.child.b -> grand.parent.child.a -> grand.parent.a)
grand/a.py:2:27: PLR0401 Cyclic import (grand.a -> grand.parent.child.b -> grand.parent.child.a)
grand/parent/a.py:2:1: I001 [*] Import block is un-sorted or un-formatted
grand/parent/a.py:2:20: PLR0401 Cyclic import (grand.parent.a -> grand.a -> grand.parent.child.a)
grand/parent/a.py:2:20: PLR0401 Cyclic import (grand.parent.a -> grand.a)
grand/parent/a.py:2:20: PLR0401 Cyclic import (grand.parent.a -> grand.a -> grand.parent.child.b -> grand.parent.child.a)
grand/parent/child/a.py:2:1: I001 [*] Import block is un-sorted or un-formatted
grand/parent/child/a.py:2:16: PLR0401 Cyclic import (grand.parent.child.a -> grand.parent.a -> grand.a)
grand/parent/child/a.py:2:16: PLR0401 Cyclic import (grand.parent.child.a -> grand.parent.a -> grand.a -> grand.parent.child.b)
grand/parent/child/a.py:2:16: PLR0401 Cyclic import (grand.parent.child.a -> grand.a)
grand/parent/child/a.py:2:16: PLR0401 Cyclic import (grand.parent.child.a -> grand.a -> grand.parent.child.b)
grand/parent/child/b.py:2:15: PLR0401 Cyclic import (grand.parent.child.b -> grand.parent.child.a -> grand.parent.a -> grand.a)
grand/parent/child/b.py:2:15: PLR0401 Cyclic import (grand.parent.child.b -> grand.parent.child.a -> grand.a)
```

**Comment by chanman3388** [comment]
One thing I want to ask, for the moment I've made it such that in each module the cyclic import is present, there's a `Diagnostic`/`Message` for each one. I could make it more efficient by not doing this, but I kind of feel like it should be reported in each one. Thoughts @charliermarsh?

**Comment by charliermarsh** [comment]
@chanman3388 - Yeah I think it's okay to flag once per file per module.

As a heads up, I might ask that we carve the actual cycle detection and rule out into a separate PR, so that they're independently reviewable and mergable.

**Comment by chanman3388** [comment]
So this time, I believe it's ready. I didn't need the module mappings in the end.
Here is the implementation of the cyclic imports rule.
https://github.com/chanman3388/ruff/pull/1
I still need to add a fixture for it.

**Comment by chanman3388** [comment]
@charliermarsh Something I wasn't sure upon, what did you mean by "The cycle-detection PR would mark this branch as its upstream" exactly? I've made a PR which merged in the cyclic import detection rule implementation in my own fork, but that naturally won't show here.
I've linked it in my previous message so others can look at it.

**Comment by chanman3388** [comment]
Another question I had was related to fixtures. I think fixtures only work on file paths rather than directories? I think we'd need more plumbing if we wanted to test this feature via the current method?

**Comment by charliermarsh** [comment]
I think you're right.

File: crates/ruff_cli/src/diagnostics.rs
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:65]: Why did this need to be boxed?

Comment (chanman3388) [reply crates/ruff_cli/src/diagnostics.rs:65]: Clippy said the type was too large and suggested boxing it. It was a warning.

**Comment by chanman3388** [comment]
@charliermarsh what else would you like see on this PR? Should we implement the feature for fixtures to also be directories separately? There some more small changes I'd like to cherry pick from the rule implementation PR. I think for now I'll have to just unit test it.

**Comment by charliermarsh** [comment]
@chanman3388 - I think this is reasonable as far as the scope of the PR goes. I just need to find a bit of time to do a final pass on it and some benchmarking before I'd be able to merge.

Heads up that I may also iterate a bit on the exact model and execution flow we use here over time as we expand the need to resolve imports. (For example, Pyright does a pass collecting and resolving all imports prior to traversing the AST for semantic analysis. Right now, we're going in the opposite order.)

**Comment by chanman3388** [comment]
> @chanman3388 - I think this is reasonable as far as the scope of the PR goes. I just need to find a bit of time to do a final pass on it and some benchmarking before I'd be able to merge.
> 
> Heads up that I may also iterate a bit on the exact model and execution flow we use here over time as we expand the need to resolve imports. (For example, Pyright does a pass collecting and resolving all imports prior to traversing the AST for semantic analysis. Right now, we're going in the opposite order.)
> 

No problem, I had some minor improvement where I used an `Rc<String>` instead of the module name. I also saw an interesting crate which we might find useful? https://github.com/xfbs/imstr

**Comment by charliermarsh** [comment]
Thanks @chanman3388 for all your work here! I made a few small touch-ups, mostly just renaming a few of the structs and moving them into new modules. Let me know if you have any questions or run into any issues as you rebase the cycle detection work.

**Pull request merged**
